File: android/app/src/main/kotlin/com/jileduo/finance/Bidding.kt
package com.jileduo.finance

import com.qq.e.comm.pi.IBidding

/** 竞价判定结果 */
enum class BidResult {
    WON,      // eCPM >= 底价，已上报竞胜，应展示
    LOST,     // eCPM < 底价，已上报竞败，应放弃展示
    SKIPPED   // eCPM<=0（非竞价位/取不到价），未上报，按现状展示
}

/** 统一竞价判定：仅做出价上报与决策，不发事件、不展示。调用方据返回值处理展示与兜底。 */
object Bidding {
    private const val LOSS_REASON_LOW_PRICE = 1 // BiddingLossReason.LOW_PRICE

    fun evaluate(ecpm: Int, floor: Int, ad: IBidding): BidResult {
        if (ecpm <= 0) return BidResult.SKIPPED
        return if (ecpm >= floor) {
            ad.sendWinNotification(mapOf<String, Any>(
                IBidding.EXPECT_COST_PRICE to ecpm,
                IBidding.HIGHEST_LOSS_PRICE to floor
            ))
            BidResult.WON
        } else {
            ad.sendLossNotification(mapOf<String, Any>(
                IBidding.WIN_PRICE to floor,
                IBidding.LOSS_REASON to LOSS_REASON_LOW_PRICE,
                IBidding.ADN_ID to AdConfig.SELF_ADN_ID
            ))
            BidResult.LOST
        }
    }
}

File: ios/Runner/Bidding.swift
import Foundation

/// 竞价判定结果
enum BidResult {
    case won      // eCPM >= 底价，已上报竞胜，应展示
    case lost     // eCPM < 底价，已上报竞败，应放弃展示
    case skipped  // eCPM<=0（非竞价位/取不到价），未上报，按现状展示
}

/// 统一竞价判定：仅做出价上报与决策，不发 Flutter 事件、不展示。
/// 调用方据返回值处理展示与事件兜底。
/// - ad: 任意遵循 GDTAdProtocol 的广告对象（开屏/激励/插屏/Banner/原生）。
func gdtEvaluateBid(eCPM: Int, floor: Int, ad: GDTAdProtocol) -> BidResult {
    if eCPM <= 0 { return .skipped }
    if eCPM >= floor {
        ad.sendWinNotification?(withInfo: [
            "expectCostPrice": NSNumber(value: eCPM),
            "highestLossPrice": NSNumber(value: floor),
        ])
        return .won
    }
    ad.sendLossNotification?(withInfo: [
        "winPrice": NSNumber(value: floor),
        "lossReason": NSNumber(value: 1),       // GDTAdBiddingLossReasonLowPrice
        "adnId": AdConfig.kSelfAdnId,
    ])
    return .lost
}
